go/libraries/doltcore/sqle/binlogreplication/binlog_json_serialization.go

```diff
@@ -0,0 +1,352 @@
+				return 0, nil, fmt.Errorf("unexpected buffer length")
+			}
+			valueEntriesBuffer = appendForEncoding(valueEntriesBuffer, uint32(encodedValue[0]), largeEncoding)
+		} else {
+			if !largeEncoding && nextValuesOffset > maxOffsetSize-uint32(len(encodedValue)) {
+				return 0, nil, fmt.Errorf("offset too large for small array encoding")
+			}
+
+			valueEntriesBuffer = append(valueEntriesBuffer, typeId)
+			valueEntriesBuffer = appendForEncoding(valueEntriesBuffer, nextValuesOffset, largeEncoding)
+			valuesBuffer = append(valuesBuffer, encodedValue...)
+			nextValuesOffset += uint32(len(encodedValue))
+		}
+	}
+
+	// element count (uint16 for small arrays)
+	encodedArray = appendForEncoding(encodedArray, uint32(len(jsonArray)), largeEncoding)
+
+	// Grab the total size of the array data from the next offset position pointing to the end of the values buffer
+	arrayPayloadLength := nextValuesOffset
+
+	encodedArray = appendForEncoding(encodedArray, arrayPayloadLength, largeEncoding)
+	encodedArray = append(encodedArray, valueEntriesBuffer...)
+	encodedArray = append(encodedArray, valuesBuffer...)
+
+	if !largeEncoding {
+		return jsonTypeSmallArray, encodedArray, nil
+	} else {
+		return jsonTypeLargeArray, encodedArray, nil
+	}
+}
+
+// encodeJsonObject encodes the specified |jsonObject| into MySQL's internal JSON encoding and returns
+// the type ID indicating whether this is a small or large object, the encoded object data, and any
+// error encountered. The |largeEncoding| param controls whether this function will use the small
+// object encoding (i.e. using 2 bytes for counts, sizes, and offsets), or the large object encoding
+// (i.e. using 4 bytes for counts, sizes, and offsets).
+//
+// A JSON Object is encoded into the following components:
+// - Type Identifier: Always 1 byte; either jsonTypeSmallObject or jsonTypeLargeObject (not included in returned []byte)
+// - Count: variable based on small/large encoding; holds the number of keys in the object
+// - Size: variable based on small/large encoding; total size of the encoded object (i.e. everything but the Type ID)
+// - Key Entries: 1 per key; variable length key offset (based on small/large encoding), plus 2 bytes for key length
+// - Value Entries (variable): 1 per value; 1 byte for type ID, 2 bytes for offset or inlined literal value for jsonTypeSmallObject, otherwise 4
+// - Keys (variable): 1 per key; encoded string bytes
+// - Values (variable): 1 per value; encoded value bytes
+func encodeJsonObject(jsonObject map[string]any, largeEncoding bool) (typeId byte, encodedObject []byte, err error) {
+	var keyEntriesBuffer []byte
+	var keysBuffer []byte
+	nextKeysOffset := calculateInitialObjectKeysOffset(len(jsonObject), largeEncoding)
+
+	// Sort the keys so that we can process the keys and values in a consistent order
+	sortedKeys := make([]string, 0, len(jsonObject))
+	for key := range jsonObject {
+		sortedKeys = append(sortedKeys, key)
+	}
+	sort.Strings(sortedKeys)
```

Comment:
Not sure if it makes any difference in replication, but I think MySQL sorts JSON keys by length then alphabetically.

Reply:
oh wow, that's a quirky sort order 😄  I don't think it matters for replication (I've got some integration tests that actually replicate all the data types to MySQL and it seems happy to get them in this order), but that's worth adding a note into the code for.